pkg/autoupdate/tuf/library_manager.go

```diff
@@ -0,0 +1,341 @@
+package tuf
+
+import (
+	"context"
+	"crypto/sha512"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/Masterminds/semver"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/kolide/kit/fsutil"
+	"github.com/kolide/kit/version"
+	"github.com/kolide/launcher/ee/localserver"
+	"github.com/kolide/launcher/pkg/autoupdate"
+	client "github.com/theupdateframework/go-tuf/client"
+)
+
+// updateLibraryManager manages the update libraries for launcher and osquery.
+// It downloads and verifies new updates, and moves them to the appropriate
+// location in the library specified by the version associated with that update.
+// It also ensures that old updates are removed when they are no longer needed.
+type updateLibraryManager struct {
+	metadataClient  *client.Client // used to validate downloads
+	mirrorUrl       string         // dl.kolide.co
+	mirrorClient    *http.Client
+	rootDirectory   string
+	operatingSystem string
+	osquerier       localserver.Querier // used to query for current running osquery version
+	logger          log.Logger
+}
+
+func newUpdateLibraryManager(metadataClient *client.Client, mirrorUrl string, mirrorClient *http.Client, rootDirectory string, operatingSystem string, osquerier localserver.Querier, logger log.Logger) (*updateLibraryManager, error) {
+	ulm := updateLibraryManager{
+		metadataClient:  metadataClient,
+		mirrorUrl:       mirrorUrl,
+		mirrorClient:    mirrorClient,
+		rootDirectory:   rootDirectory,
+		operatingSystem: operatingSystem,
+		osquerier:       osquerier,
+		logger:          log.With(logger, "component", "tuf_autoupdater_library_manager"),
+	}
+
+	// Ensure our staged updates and updates directories exist
+	for _, binary := range binaries {
+		// Attempt out the staged updates directory, in case there are any archives there that we
+		// were unable to clean up previously.
+		_ = os.RemoveAll(ulm.stagedUpdatesDirectory(binary))
+
+		// Create the directory for staging updates
+		if err := os.MkdirAll(ulm.stagedUpdatesDirectory(binary), 0755); err != nil {
+			return nil, fmt.Errorf("could not make staged updates directory for %s: %w", binary, err)
+		}
+
+		// Create the update library
+		if err := os.MkdirAll(ulm.updatesDirectory(binary), 0755); err != nil {
+			return nil, fmt.Errorf("could not make updates directory for %s: %w", binary, err)
+		}
+	}
+
+	return &ulm, nil
+}
+
+// updatesDirectory returns the update library location for the given binary.
+func (ulm *updateLibraryManager) updatesDirectory(binary string) string {
+	return filepath.Join(ulm.rootDirectory, fmt.Sprintf("%s-updates", binary))
```

Comment:
seems okay. Maybe a slight preference for `<basedir>/<binary>` which would commonly expand to `<launcher root>/updates/<binary>`

Reply:
Sure, I like that 👍

Reply:
Implied here, is that we need to be able to configure what `basedir`, but it would default to root/updates

Reply:
I've updated this!